juju1/state/migration_export.go

```diff
@@ -931,6 +944,76 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 	return nil
 }
 
+func extraBindings(application *Service) ([]string, error) {
+	// Collect any extra bindings from the charm. These aren't
+	// captured in the db but they're stored in the charm blob, in the
+	// metadata file.
+	ch, _, err := application.Charm()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	st := application.st
+	store := statestorage.NewStorage(st.EnvironUUID(), st.MongoSession())
+	reader, _, err := store.Get(ch.StoragePath())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer reader.Close()
+
+	localFile, err := ioutil.TempFile("", "charm-"+ch.URL().Name)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer os.Remove(localFile.Name())
+	defer localFile.Close()
+
+	numBytes, err := io.Copy(localFile, reader)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zipFile, err := zip.NewReader(localFile, numBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Find metadata.yaml.
+	var metadataFile *zip.File
+	for _, f := range zipFile.File {
+		if f.Name == "metadata.yaml" {
+			metadataFile = f
+		}
+	}
+	if metadataFile == nil {
+		return nil, errors.Errorf("no metadata file in the %q charm!", ch.Meta().Name)
```

Comment:
Should this be fatal or just a warning? do we really care to halt the whole operation if there's a single dodgy charm? there'd be do extra bindings in that case anyway. Same comment to some of the other errors below, eg the yaml one

Reply:
Yeah, good call - I'll change that.